miniapps/nurbs/nurbs_ex25p.cpp

```diff
@@ -0,0 +1,804 @@
+   //    when eliminating the non-homogeneous boundary condition to modify the
+   //    r.h.s. vector b.
+   ParComplexGridFunction x(fespace);
+   x = 0.0;
+   VectorFunctionCoefficient E_Re(dim, E_bdr_data_Re);
+   VectorFunctionCoefficient E_Im(dim, E_bdr_data_Im);
+
+   // 9. Set up the bilinear form corresponding to the EM diffusion operator
+   //    curl muinv curl + sigma I, by adding the curl-curl and the mass domain
+   //    integrators.
+   Array<int> attr;
+   Array<int> attrPML;
+   if (pmesh->attributes.Size())
+   {
+      attr.SetSize(pmesh->attributes.Max());
+      attrPML.SetSize(pmesh->attributes.Max());
+      attr = 0;   attr[0] = 1;
+      attrPML = 0;
+      if (pmesh->attributes.Max() > 1)
+      {
+         attrPML[1] = 1;
+      }
+   }
+
+   ConstantCoefficient muinv(1_r / mu);
+   ConstantCoefficient omeg(-pow2(omega) * epsilon);
+   RestrictedCoefficient restr_muinv(muinv,attr);
+   RestrictedCoefficient restr_omeg(omeg,attr);
+
+   // Integrators inside the computational domain (excluding the PML region)
+   ParSesquilinearForm a(fespace, conv);
+   a.AddDomainIntegrator(new CurlCurlIntegrator(restr_muinv),NULL);
+   a.AddDomainIntegrator(new VectorFEMassIntegrator(restr_omeg),NULL);
+
+   int cdim = (dim == 2) ? 1 : dim;
+   PMLDiagMatrixCoefficient pml_c1_Re(cdim,detJ_inv_JT_J_Re, pml);
+   PMLDiagMatrixCoefficient pml_c1_Im(cdim,detJ_inv_JT_J_Im, pml);
+   ScalarVectorProductCoefficient c1_Re(muinv,pml_c1_Re);
+   ScalarVectorProductCoefficient c1_Im(muinv,pml_c1_Im);
+   VectorRestrictedCoefficient restr_c1_Re(c1_Re,attrPML);
+   VectorRestrictedCoefficient restr_c1_Im(c1_Im,attrPML);
+
+   PMLDiagMatrixCoefficient pml_c2_Re(dim, detJ_JT_J_inv_Re,pml);
+   PMLDiagMatrixCoefficient pml_c2_Im(dim, detJ_JT_J_inv_Im,pml);
+   ScalarVectorProductCoefficient c2_Re(omeg,pml_c2_Re);
+   ScalarVectorProductCoefficient c2_Im(omeg,pml_c2_Im);
+   VectorRestrictedCoefficient restr_c2_Re(c2_Re,attrPML);
+   VectorRestrictedCoefficient restr_c2_Im(c2_Im,attrPML);
+
+   // Integrators inside the PML region
+   a.AddDomainIntegrator(new CurlCurlIntegrator(restr_c1_Re),
+                         new CurlCurlIntegrator(restr_c1_Im));
+   a.AddDomainIntegrator(new VectorFEMassIntegrator(restr_c2_Re),
+                         new VectorFEMassIntegrator(restr_c2_Im));
+
+   // 10. Assemble the bilinear form and the corresponding linear system,
+   //     applying any necessary transformations such as: eliminating boundary
+   //     conditions, applying conforming constraints for non-conforming AMR,
+   //     static condensation, etc.
+
+   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+   a.Assemble(0);
+
+   OperatorPtr A;
+   Vector B, X;
+   a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+
+   cout << "Size of linear system: " << A->Height() << endl;
+
+   ConstantCoefficient absomeg(pow2(omega) * epsilon);
+   RestrictedCoefficient restr_absomeg(absomeg,attr);
+
+   ParBilinearForm prec(fespace);
```

Comment:
It would be helpful to add some comments about the preconditioner assembled here.